Fix parallel `make check`

rprichard opened the pull request:
This required fixing the `pretty-rpass-full` tests to have the same `$$(CSREQ$(1)_T_$(2)_H_$(3))`  dependencies as the `rpass-full` and `cfail-full` tests.  It also required fixing the `run-make/simd-ffi` test to use unique names for its output files.

rust-highfive commented:
r? @brson

(rust_highfive has picked a reviewer for you, use r? to override)

rprichard commented:
cc @pnkfelix 

So far, I've tested this PR using `make -j10 check`, in a clean build directory.

This PR fixes #22021.

alexcrichton commented:
I think there's still a problem if `run-pass` and `pretty-run-pass` run concurrently as they'll tromp over each other when building the `// aux-build` directives, but other than that this is all I know of!

brson commented:
@bors r+ rollup

Nice fixes!

bors commented:
:pushpin: Commit 566d653 has been approved by `brson`

<!-- @bors r=brson 566d6531fca08a9645dfef6c9ea96477d87a461c -->

rprichard commented:
> I think there's still a problem if run-pass and pretty-run-pass run concurrently as they'll tromp over each other when building the // aux-build directives, but other than that this is all I know of!

Thanks for letting me know.  If I have time, I'll look into it.

bors commented:
:hourglass: Testing commit 566d653 with merge 9da3545...

bors commented:
:broken_heart: Test failed - [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/4529)

steveklabnik commented:
@bors: r-

steveklabnik commented:
This seems to be causing the rollup to fail, due to android. Any ideas?

rprichard commented:
I would think that fulldeps tests shouldn't run on cross-compiled targets, because our build system doesn't compile LLVM for those targets.  I'll look into it.

rprichard commented:
The fulldeps tests require the rustc crates to be built, but there are two kinds of fulldeps tests:

 * Plugin tests that need host crates
 * Non-plugin tests that need target crates

When I submitted this PR, I was only thinking about the non-plugin tests.  In fact, for the `rpass-full` and `cfail-full` tests, we only register the `CSREQ` dependency for the host crates (for syntax plugins):
```
CTEST_DEPS_rpass-full_$(1)-T-$(2)-H-$(3) = $$(RPASS_FULL_TESTS) $$(CSREQ$(1)_T_$(3)_H_$(3)) $$(SREQ$(1)_T_$(2)_H_$(3))
CTEST_DEPS_cfail-full_$(1)-T-$(2)-H-$(3) = $$(CFAIL_FULL_TESTS) $$(CSREQ$(1)_T_$(3)_H_$(3)) $$(SREQ$(1)_T_$(2)_H_$(3))
```
We depend upon `$$(CSREQ$(1)_T_$(3)_H_$(3))`.  The `T_$(3)` is correct for plugin tests, but wrong for non-plugin tests.  We can't just add the `T_$(2)` dependency, though, because `librustllvm.a` isn't always built for target triples (as demonstrated in the Android failure in this PR).

Currently, if someone attempts to run `make check` in a configuration with a target-only triple, the non-plugin fulldeps tests will generally fail.  If that target-only triple is Android, though, it will succeed, because we've skipped Android on the non-plugin tests.

The immediate fix for this PR is to do the same thing with the pretty fulldeps tests that we already do with the other fulldeps tests.

At one point, we were skipping fulldeps tests for target-only triples, using the `CTEST_DISABLE_NONSELFHOST` mechanism.  The mechanism is still there, but it's not used anymore.  Maybe it should be reenabled, but only for non-plugin tests, somehow.

alexcrichton commented:
I think it should also be possible to just skip fulldeps tests whenever the target is not equal to the host as those can't link to the host libraries anyway (as they probably weren't built).

rprichard commented:
That's definitely possible -- it was implemented in [this commit](https://github.com/rust-lang/rust/commit/4edf7b8c3478365a3c327d7fe340447f9d328d37) on April 2 last year, then disabled in [this commit](https://github.com/rust-lang/rust/commit/4ac89cd27613ffe574a6fc6009793c1900e5cfca) two weeks later.

The problem with the idea is: for cross-compiles, about half of the fulldeps tests are verifying that `rustc` correctly searches for a compiler plugin of the *host* triple, not the *target* triple. If you search [this Android build log](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/4528/steps/test/logs/stdio) for `run-pass-fulldeps`, you'll see that about half of the tests are ignored (because there's no Android-targeted LLVM), but the other half pass.

I did notice that the second commit I linked removed a bunch of `// ignore-cross-compile` lines.  I bet we could replace the existing `// ignore-android` lines, and it would fix the general cross-compilation issue.

alexcrichton commented:
Hm I think that the removal of `ignore-cross-compile` was intentional at that time (as the change enabled the tests to pass), but I imagine that changing many `ignore-android` to `ignore-cross-compile` would be much more "semantically correct" at least!

rprichard commented:
Yes, the removing of the `ignore-cross-compile` was intentional, because the syntax extension tests work fine with cross-compilation, which the buildbot tests today for Android.  At the time it was removed, the quoting run-pass tests were the only fulldeps tests that broke with cross-compilation, and they were marked `ignore-android` or `ignore-test`.

I'm trying to change the fulldeps `ignore-android` directives to `ignore-cross-compile` and testing it with a cross-compile from x86_64 to i686.  This is easier to setup than an Android cross-compile.  The first thing I've noticed is that rustdoc tests using external crates also fail, and they too are marked `ignore-android`, so I'm changing those also.

alexcrichton commented:
Sounds good to me, thanks @rprichard!

rprichard commented:
I added a couple of commits that switch over to using `ignore-cross-compile`.  It has passed `make check` on an x86_64 machine with an extra i686 target.  I haven't tried testing Android, because it's inconvenient.  It also didn't pass `make -j10 check`, ironically, because I think I finally hit the "run-pass and pretty-run-pass run concurrently" problem that @alexcrichton mentioned originally.

tamird commented on src/test/run-pass-fulldeps/qquote.rs:
```diff
@@ -8,6 +8,7 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+// ignore-cross-compile
```
did this one pass previously?

rprichard replied:
I believe the `// ignore-test` line in that file means that the test is unconditionally skipped.  I verified the the test is skipped on one of the Linux buildbot builders.  In any case, it needs `// ignore-cross-compile` because it uses the `syntax` crate.

tamird replied:
ah, I see. I'm trying to get this test compiling, but it's pretty busted :(

rprichard commented:
So, the appropriate title of this PR is now: "Fix cross-compiling `make check`, and get parallel `make check` incrementally closer to working".  We could merge it as-is, or I could see how hard it is to fix the pretty vs non-pretty aux-build problem.

alexcrichton commented:
@bors: r+ 528472e651257bc07d8a77d25a0948436bada57a

rprichard commented:
Adding the `config.mode` to the `aux_output_dir_name()` in `compiletest` should fix the known remaining problem with parallel `make check`, I think.  Can I add a new commit to this PR (given that it has already been approved)?  I assume I still need to test it, which will take a bit of time.

It will make paths a bit longer, which shouldn't be a problem on Unix, but Windows might be a concern, given its 260-character limit.  The mode is frequently redundant with the directory name of the test source file, but omitting it seems hazardous.  e.g. `debuginfo-lldb` and `debuginfo-gdb` use the same input files, and in principle, they could run concurrently, although currently, only one can be enabled.

alexcrichton commented:
Feel free to throw it in, that sounds like a good fix!

rprichard commented:
I think I should also squash my two commits touching `tests.mk`.  The first commit is wrong, and the second commit corrects it.

rprichard commented:
I squashed my two `tests.mk` changes and added a new commit changing `compiletest`.  (Regarding the `compiletest` commit, previously we had been suffixing `.libaux` but somewhere the `.` was lost.  I assumed it was an accident and put the `.` back.)

aux dirs now look like this:
```
./x86_64-unknown-linux-gnu/test/run-pass/check-static-recursion-foreign.stage2-i686-unknown-linux-gnu.pretty.libaux
./x86_64-unknown-linux-gnu/test/run-pass/check-static-recursion-foreign.stage2-i686-unknown-linux-gnu.run-pass.libaux
```

alexcrichton commented:
@bors: r+ 38d26d8

bors commented:
:hourglass: Testing commit 38d26d8 with merge 69e47c7...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/4640), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/4647), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/4609), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/4628), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/4596), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/4661), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/4660), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/4639), [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/4329), [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/4315), [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/3132), [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/3152)
